server/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -310,6 +337,10 @@ protected void doRun() throws Exception {
         @Override
         public void onResponse(PrimaryShardReference primaryShardReference) {
             try {
+                final ClusterState clusterState = clusterService.state();
+                if (handleBlockExceptions(clusterState, request, this::handleBlockException)) {
```

Comment:
handleBlockException re-resolves the index. I think we should always use the PrimaryShardReference here to make sure we always use the right one (with uuid and all).

Reply:
Good catch